test/affinity-balancer.test.ts

```diff
@@ -0,0 +1,357 @@
+      } as any)
+    );
+  }
+
+  const results = await Promise.all(taskPromises);
+  assert.deepStrictEqual(results, [4, 4, 4, 4, 4], 'All tasks should complete');
+
+  await pool.close();
+});
+
+test('AffinityBalancer: mixed affinity and non-affinity tasks', async () => {
+  const pool = new Piscina({
+    filename: resolve(__dirname, 'fixtures/eval.js'),
+    maxThreads: 3,
+    minThreads: 3,
+    concurrentTasksPerWorker: 5,
+    loadBalancer: Piscina.AffinityBalancer({ maximumUsage: 5 })
+  });
+
+  const taskPromises = [];
+  
+  // Mix affinity and non-affinity tasks
+  for (let i = 0; i < 3; i++) {
+    // Affinity task
+    taskPromises.push(
+      pool.run('2 + 2', {
+        [Piscina.queueOptionsSymbol]: { affinityKey: 1 }
+      } as any)
+    );
+    // Non-affinity task
+    taskPromises.push(pool.run('3 + 3'));
+  }
+
+  const results = await Promise.all(taskPromises);
+  assert.strictEqual(results.length, 6, 'All tasks should complete');
+  assert.strictEqual(results[0], 4, 'First affinity task');
+  assert.strictEqual(results[1], 6, 'First non-affinity task');
+
+  await pool.close();
+});
+
+test('AffinityBalancer: prefers idle worker when available', async () => {
+  const pool = new Piscina({
+    filename: resolve(__dirname, 'fixtures/eval.js'),
+    maxThreads: 3,
+    minThreads: 3,
+    concurrentTasksPerWorker: 10,
+    loadBalancer: Piscina.AffinityBalancer({ maximumUsage: 10 })
+  });
+
+  const taskPromises = [];
+  
+  // First task with key 1 - will go to some worker
+  taskPromises.push(
+    pool.run('2 + 2', {
+      [Piscina.queueOptionsSymbol]: { affinityKey: 1 }
+    } as any)
+  );
+
+  // Wait for first task to complete
+  await taskPromises[0];
+
+  // Second task with same key should go back to same worker (now idle)
+  taskPromises.push(
+    pool.run('3 + 3', {
+      [Piscina.queueOptionsSymbol]: { affinityKey: 1 }
+    } as any)
+  );
+
+  const results = await Promise.all(taskPromises);
+  assert.deepStrictEqual(results, [4, 6], 'All tasks should complete');
+
+  await pool.close();
+});
+
+test('AffinityBalancer: handles concurrent tasks with same affinity key', async () => {
+  const pool = new Piscina({
+    filename: resolve(__dirname, 'fixtures/eval.js'),
+    maxThreads: 2,
+    minThreads: 2,
+    concurrentTasksPerWorker: 3,
+    loadBalancer: Piscina.AffinityBalancer({ maximumUsage: 3 })
+  });
+
+  const taskPromises = [];
+  
+  // Many concurrent tasks with same key - should queue on the preferred worker
+  for (let i = 0; i < 6; i++) {
+    taskPromises.push(
+      pool.run('new Promise(resolve => setTimeout(() => resolve(42), 10))', {
+        [Piscina.queueOptionsSymbol]: { affinityKey: 1 }
+      } as any)
+    );
+  }
+
+  const results = await Promise.all(taskPromises);
+  assert.strictEqual(results.length, 6, 'All tasks should complete');
+  assert.ok(results.every(r => r === 42), 'All results should be 42');
+
+  await pool.close();
+});
```

Comment:
These tests are not testing the tasks are being routed accordingly to the affinity key